### PR TITLE
SDL Core implementation of feature - 0261 New vehicle data WindowStatus

### LINKED ISSUE #3193: [SDL 0261] New vehicle data WindowStatus
Proposal: [New vehicle data WindowStatus](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0261-New-vehicle-data-WindowStatus.md)

> This proposal is to add new vehicle data `WindowStatus` to know the position of windows in a vehicle.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/869

Steering Committee Decision:

>The Steering Committee voted to accept this proposal with the revision to `WindowStatus` described in [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/869#issuecomment-564050172).

> The proposal .md file was updated to reflect these revisions on 12/12/19.

### PR DESCRIPTION
**Pull request by YarikMamykin**
Implements/Fixes #3193 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
Unit testing + ATF testing 

### Summary
1. Added new `WindowStatus` and `WindowState` data types to API.
2. Updated VehicleData RPCs with new parameters (Subscribe/Unsubscribe/Get/OnVehicleData).
3. Updated unit tests.
4. Updated `sdl_preloaded_pt.json` file.

### CLA
- [X] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/appMain/sdl_preloaded_pt.json [hunk 1]
```diff
@@ -3914,6 +3918,30 @@
                     "mandatory": false,
                     "since": "5.1"
                 },
+                {
```
Comment (IGapchuk) [inline src/appMain/sdl_preloaded_pt.json:3921]: @IGapchuk move to the separate commit

Comment (YarikMamykin) [reply src/appMain/sdl_preloaded_pt.json:3921]: @IGapchuk all changes already are in separate commit 15cfd4b. Why should I do a separate commit for this particular change?

File: src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+  application_manager::AppExtensionPtr vi_app_extension_ptr_;
+  std::shared_ptr<sync_primitives::Lock> app_set_lock_ptr_;
+  vehicle_info_plugin::VehicleInfoPlugin vi_plugin_;
+  application_manager_test::MockRPCHandler mock_rpc_handler_;
+  resumption::LastStateWrapperPtr mock_last_state_;
+};
+
+TEST_F(SubscribeVehicleRequestTest, Run_AppNotRegistered_UNSUCCESS) {
+  CommandPtr command(CreateCommandVI<SubscribeVehicleDataRequest>());
+  EXPECT_CALL(app_mngr_, application(_))
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc:118]: Please check application id

File: src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+}
+
+TEST_F(SubscribeVehicleRequestTest, Run_DataToSubscribeIsNotExisted_UNSUCCESS) {
+  MessageSharedPtr command_msg(CreateMessage(smart_objects::SmartType_Map));
+  (*command_msg)[am::strings::params][am::strings::connection_key] =
+      kConnectionKey;
+  (*command_msg)[am::strings::msg_params][am::strings::button_name] =
+      kVehicleTypeStr;
+
+  am::VehicleData data;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc:137]: Please rename the variable to empty_data for better understanding

File: src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+  EXPECT_CALL(mock_message_helper_, vehicle_data())
+      .WillRepeatedly(ReturnRef(data));
+
+  smart_objects::SmartObject empty_hmi_custom_params;
+  ON_CALL(mock_custom_vehicle_data_manager_, CreateHMIMessageParams(_))
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc:142]: Should we check the parameter?

File: src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+  MessageSharedPtr command_msg(CreateMessage(smart_objects::SmartType_Map));
+  (*command_msg)[am::strings::params][am::strings::connection_key] =
+      kConnectionKey;
+  (*command_msg)[am::strings::msg_params][kMsgParamKey] = false;
+  CommandPtr command(CreateCommandVI<SubscribeVehicleDataRequest>(command_msg));
+  ASSERT_TRUE(command->Init());
+  command->Run();
+}
+
+TEST_F(SubscribeVehicleRequestTest, Run_SubscribeDataDisabled_UNSUCCESS) {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc:184]: Can't see the place in UT, where data is diabled. Make iot more obvious

File: src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+  (*command_msg)[am::strings::msg_params]
+                [application_manager::strings::window_status] = true;
+
+  CommandPtr command(CreateCommandVI<SubscribeVehicleDataRequest>(command_msg));
+
+  EXPECT_CALL(app_mngr_, application(kConnectionKey))
+      .WillRepeatedly(Return(mock_app_));
+  EXPECT_CALL(app_mngr_, hmi_interfaces())
+      .WillRepeatedly(ReturnRef(mock_hmi_interfaces_));
+  EXPECT_CALL(mock_hmi_interfaces_, GetInterfaceState(_))
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc:223]: Interface name?

Comment (YarikMamykin) [reply src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc]: @LuxoftAKutsan interface name check added in https://github.com/smartdevicelink/sdl_core/pull/3391/commits/c32519a9280675d19886de786afe15f756f0de31

File: src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc (hunk 3 above)
Comment (YarikMamykin) [reply src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc]: @LuxoftAKutsan fixed in https://github.com/smartdevicelink/sdl_core/pull/3391/commits/c32519a9280675d19886de786afe15f756f0de31

File: src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc (hunk 2 above)
Comment (YarikMamykin) [reply src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc]: @LuxoftAKutsan check added in https://github.com/smartdevicelink/sdl_core/pull/3391/commits/c32519a9280675d19886de786afe15f756f0de31

File: src/appMain/sdl_preloaded_pt.json (hunk 1 above)
Comment (YarikMamykin) [reply src/appMain/sdl_preloaded_pt.json:3921]: @IGapchuk section moved to commit https://github.com/smartdevicelink/sdl_core/pull/3391/commits/f37eef4b7a7043bcf7e11b7860d41b54a7ec2b4b

File: src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc (hunk 5 above)
Comment (YarikMamykin) [reply src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc]: @LuxoftAKutsan renamed UT so it would better describe its logic. https://github.com/smartdevicelink/sdl_core/pull/3391/commits/2856a9cfbfd40b464cb9bf80a2fdec13dd716f96

File: src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc (hunk 4 above)
Comment (YarikMamykin) [reply src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc]: @LuxoftAKutsan check added https://github.com/smartdevicelink/sdl_core/pull/3391/commits/2856a9cfbfd40b464cb9bf80a2fdec13dd716f96

**Comment by LitvinenkoIra** [comment]
@atiwari9 please review this PR

**Comment by atiwari9** [comment]
PR Approved.

Verified with ATF scripts commit: [623f2eb](https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2411/commits/623f2eb3c4f50032a37c7245f854aced7678abcb)

Logs: 
[Test_Report_WindowStatus_623f2eb.zip](https://github.com/smartdevicelink/sdl_atf_test_scripts/files/4793189/Test_Report_WindowStatus_623f2eb.zip)

**Comment by LitvinenkoIra** [comment]
@theresalech PR is ready for Livio review. Thank you!

File: src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+ LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdint.h>
```
Comment (ShobhitAd) [inline src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc:33]: Could you please explain why this include is required. If it is not required, I think it should be removed

Comment (LitvinenkoIra) [reply src/components/application_manager/rpc_plugins/vehicle_info_plugin/test/commands/mobile/subscribe_vehicle_request_test.cc]: @ShobhitAd removed in 2aa0260

**Comment by LitvinenkoIra** [comment]
@ShobhitAd Thank you for the review, please check the latest updates

File: src/appMain/sdl_preloaded_pt.json [hunk 8]
```diff
@@ -3914,6 +3918,30 @@
...
+                    "name": "windowStatus",
+                    "params": [
+                        {
+                            "name": "location",
+                            "key": "OEM_REF_WND_LOC",
+                            "type": "Grid",
+                            "mandatory": true
+                        },
```
Comment (ShobhitAd) [inline src/appMain/sdl_preloaded_pt.json:3929]: Referring back to this comment https://github.com/smartdevicelink/sdl_core/pull/3304#discussion_r449029139. The type can only be an enum or a base type. Grid is a Struct so it should be expanded here.
```suggestion
                        {
                            "name": "location",
                            "params": [
                                {
                                    "name": "col",
                                    "key": "OEM_REF_WND_LOC_COL",
                                    "type": "Integer",
                                    "mandatory": true,
                                    "minvalue": -1,
                                    "maxvalue": 100
                                },
                                {
                                    "name": "row",
                                    "key": "OEM_REF_WND_LOC_ROW",
                                    "type": "Integer",
                                    "mandatory": true,
                                    "minvalue": -1,
                                    "maxvalue": 100
                                },
                                {
                                    "name": "level",
                                    "key": "OEM_REF_WND_LOC_LVL",
                                    "type": "Integer",
                                    "mandatory": false,
                                    "defvalue" : 0,
                                    "minvalue": -1,
                                    "maxvalue": 100
                                },
                                {
                                    "name": "colspan",
                                    "key": "OEM_REF_WND_LOC_COLSPN",
                                    "type": "Integer",
                                    "mandatory": false,
                                    "defvalue" : 1,
                                    "minvalue": 1,
                                    "maxvalue": 100
                                },
                                {
                                    "name": "rowspan",
                                    "key": "OEM_REF_WND_LOC_ROWSPN",
                                    "type": "Integer",
                                    "mandatory": false,
                                    "defvalue" : 1,
                                    "minvalue": 1,
                                    "maxvalue": 100
                                },
                                {
                                    "name": "levelspan",
                                    "key": "OEM_REF_WND_LOC_LVLSPN",
                                    "type": "Integer",
                                    "mandatory": false,
                                    "defvalue" : 1,
                                    "minvalue": 1,
                                    "maxvalue": 100
                                }
                            ],
                            "key": "OEM_REF_WND_LOC",
                            "type": "Struct",
                            "mandatory": true
                        },
```

File: src/appMain/sdl_preloaded_pt.json [hunk 9]
```diff
@@ -3914,6 +3918,30 @@
...
+                        {
+                            "name": "state",
+                            "key": "OEM_REF_WND_STATE",
+                            "type": "WindowState",
+                            "mandatory": true
+                        }
```
Comment (ShobhitAd) [inline src/appMain/sdl_preloaded_pt.json:3935]: Referring back to this comment https://github.com/smartdevicelink/sdl_core/pull/3304#discussion_r449029139. The type can only be an enum or a base type. WindowState is a Struct so it should be expanded here.
```suggestion
                        {
                            "name": "state",
                            "params": [
                                {
                                    "name": "approximatePosition",
                                    "key": "OEM_REF_WND_STATE_APRPOS",
                                    "type": "Integer",
                                    "mandatory": true,
                                    "minvalue": 0,
                                    "maxvalue": 100
                                },
                                {
                                    "name": "deviation",
                                    "key": "OEM_REF_WND_STATE_DEV",
                                    "type": "Integer",
                                    "mandatory": true,
                                    "minvalue": 0,
                                    "maxvalue": 100
                                }
                            ],
                            "key": "OEM_REF_WND_STATE",
                            "type": "Struct",
                            "mandatory": true
                        }
```

Comment (LitvinenkoIra) [reply src/appMain/sdl_preloaded_pt.json]: @ShobhitAd Thank you for the review, all comments are processed, please check the latest updates.

**Comment by ShobhitAd** [comment]
The RPC spec PR for this feature https://github.com/smartdevicelink/rpc_spec/pull/250 has been merged into develop. Please update the `.gitmodules` file to use the `smartdevicelink` remote and update the `tools/rpc_spec` submodule commit.

**Comment by AKalinich-Luxoft** [comment]
> The RPC spec PR for this feature [smartdevicelink/rpc_spec#250](https://github.com/smartdevicelink/rpc_spec/pull/250) has been merged into develop. Please update the `.gitmodules` file to use the `smartdevicelink` remote and update the `tools/rpc_spec` submodule commit.

@ShobhitAd done in https://github.com/smartdevicelink/sdl_core/pull/3391/commits/0fd97914c6207722c2825676bae14f887ea2c1aa
